doc: add doc for infernecemodelrewrites. (#1978)
kubernetes-sigs/gateway-api-inference-extension: site-src/guides/adapter-rollout.md

without altering the underlying base model or infrastructure.
Use LoRA adapter rollouts to test improvements, bug fixes, or new features in your LoRA adapters.

The [`InferenceModelRewrite`](/api-types/inferencemodelrewrite) resource allows platform administrators and model owners to control how inference requests are routed to specific models within an Inference Pool.
This capability is essential for managing model lifecycles without disrupting client applications.

## Prerequisites & Setup

Follow [getting-started](https://gateway-api-inference-extension.sigs.k8s.io/guides/getting-started-latest/#getting-started-with-an-inference-gateway) to set up the IGW stack.

In this guide, we modify the LoRA adapters configMap to have two food-review models to better illustrate the gradual rollout scenario.

> The configMap used in this guide is as follows:

Reviewer: ```suggestion
The ConfigMap used in this guide is as follows:
```

  "model": "food-review-v1",
  "object": "chat.completion",
  "prompt_logprobs": null,
  "usage": {
    "completion_tokens": 10,
    "prompt_tokens": 43,
    "prompt_tokens_details": null,
    "total_tokens": 53
  }
}
```

## Step 2: Gradual Rollout

Now that `food-review-v2` is loaded (from the Prerequisites step), we can begin splitting traffic. Traffic splitting allows you to divide incoming traffic for a single model name across different adapters. This is critical for A/B testing or gradual updates.
You want to direct 90% of `food-review` traffic to the stable `food-review-v1` and 10% to the new `food-review-v2`.

#### InferenceModelRewrites (90 / 10 split)

Update the existing `InferenceModelRewrite`:

```yaml
apiVersion: inference.networking.x-k8s.io/v1alpha2
kind: InferenceModelRewrite
metadata:
  name: food-review-rewrite
spec:
  poolRef:
    group: inference.networking.k8s.io
    name: vllm-llama3-8b-instruct
  rules:
    - matches:
        - model:
            type: Exact
            value: food-review
      targets:
        - modelRewrite: "food-review-v1"
          weight: 90
        - modelRewrite: "food-review-v2"
          weight: 10
```

> ##### Result

Reviewer: ```suggestion
Run the test traffic script as follows:
```

Author: updated, ty!

```bash
❯ ./test-traffic-splitting.sh
---
Traffic Split Results, total requests: 20
food-review-v1: 17 requests
food-review-v2: 3 requests
```

#### InferenceModelRewrites (50 / 50 split)

To increase traffic to the new model, simply adjust the weights.

```yaml
      targets:
        - modelRewrite: "food-review-v1"
          weight: 50
        - modelRewrite: "food-review-v2"
          weight: 50
```

##### Result

```bash
❯ ./test-traffic-splitting.sh
___
Traffic Split Results, total requests: 20
food-review-v1: 10 requests
food-review-v2: 10 requests
```

> #### InferenceModelRewrites (0 / 100 split)

Reviewer: ditto

Once the new model is verified, shift all traffic to it.

```yaml
      targets:
        - modelRewrite: "food-review-v2"
          weight: 100
```

> ##### Result

Reviewer: ```suggestion
Run the test traffic script one last time:
```

      defaultBaseModel: meta-llama/Llama-3.1-8B-Instruct
      ensureExist:
        models:
        - id: food-review-v1
          source: Kawon/llama3.1-food-finetune_v14_r8
        - id: food-review-v2
          source: Kawon/llama3.1-food-finetune_v14_r8
```

**Verify Available Models**: You can query the `/v1/models` endpoint to confirm the adapters are loaded:

```bash
curl http://${IP}/v1/models | jq . 
```

## Step 1: Establish Baseline (Alias v1)

First, we establish a stable baseline where all requests for `food-review` are served by the existing version, `food-review-v1`. This decouples the client's request (for "food-review") from the specific version running on the backend.

> ### Scenario

Reviewer: I feel we don't need the header, just make it part of the intro section above

A client requests the model `food-review`. We want to ensure this maps strictly to `food-review-v1`.

> ### InferenceModelRewrite

Reviewer: header 4: ####

    - matches:
        - model:
            type: Exact
            value: food-review
      targets:
        - modelRewrite: "food-review-v1"
```

### Result

When a client requests `"model": "food-review"`, the system serves the request using `food-review-v1`.

```bash
curl http://${IP}/v1/chat/completions \
  -H "Content-Type: application/json" \
  -d 
'{ 
"model": "food-review",
"messages": [
  {
    "role": "user",
    "content": "Give me a spicy food challenge list."
  }
],
"max_completion_tokens": 10
}' | jq . 
```

Response:
```json
{
  "choices": [
    {
      "finish_reason": "length",
      "index": 0,
      "logprobs": null,
      "message": {
        "content": "Here's a list of spicy foods that can help",
        "reasoning_content": null,
        "role": "assistant",
        "tool_calls": []
      },
      "stop_reason": null
    }
  ],
  "created": 1764786158,
  "id": "chatcmpl-b10d939f-39bc-41ba-85c0-fe9b9d1ed3d9",
  "model": "food-review-v1",
  "object": "chat.completion",
  "prompt_logprobs": null,
  "usage": {
    "completion_tokens": 10,
    "prompt_tokens": 43,
    "prompt_tokens_details": null,
    "total_tokens": 53
  }
}
```

## Step 2: Gradual Rollout

> Now that `food-review-v2` is loaded (from the Prerequisites step), we can begin splitting traffic. Traffic splitting allows you to divide incoming traffic for a single model name across multiple backend models. This is critical for A/B testing or gradual updates.

Reviewer: ```suggestion
Now that `food-review-v2` is loaded (from the Prerequisites step), we can begin splitting traffic. Traffic splitting allows you to divide incoming traffic for a single model name across different adapters. This is critical for A/B testing or gradual updates.
```

> ### Scenario: 90/10 Split

Author: Updated the new structure looks like this

<img width="768" height="1486" alt="Image" src="https://github.com/user-attachments/assets/8af374b9-2738-4ce9-831c-33a540a6b5aa" />
